mesonbuild/interpreter/interpreter.py

```diff
@@ -687,10 +687,19 @@ def func_declare_dependency(self, node, args, kwargs):
         for d in deps:
             if not isinstance(d, dependencies.Dependency):
                 raise InterpreterException('Invalid dependency')
+
+        dep_err = 'declare_dependency: Entries in "{}" may only be self-built targets, external dependencies (including libraries) must go to "dependencies".'
+
         for l in libs:
             if isinstance(l, dependencies.Dependency):
-                raise InterpreterException('''Entries in "link_with" may only be self-built targets,
-external dependencies (including libraries) must go to "dependencies".''')
+                raise InvalidArguments(dep_err.format('link_with'))
+        for l in libs_whole:
```

Comment:
Despite the commit message, it seems we're also checking that anything we disallowed in link_with is also disallowed in link_whole, which is a whole 'nother failure mode we're catching now.

I'd say this is worth mentioning in the commit message.